cpp/src/arrow/util/decimal.cc

```diff
@@ -528,8 +528,8 @@ Status Decimal128::FromString(const util::string_view& s, Decimal128* out,
   int32_t parsed_scale = 0;
   if (dec.has_exponent) {
     auto adjusted_exponent = dec.exponent;
-    auto len = static_cast<int32_t>(significant_digits);
-    parsed_scale = -adjusted_exponent + len - 1;
+    parsed_scale =
+        -adjusted_exponent + static_cast<int32_t>(dec.fractional_digits.size());
```

Comment:
For `12.34e-3`, old code calcs scale = 6, now it's 5. I think 5 is correct.
Just curious, looks the final precision = 4 < scale = 6, is it okay?

Reply:
Precision is computed as the number of significant digits. I think it's ok (it was also the intent of the code previously).